text/3488-deprecate-cargo-categories.md

```diff
@@ -0,0 +1,397 @@
+
+Finally, having the community decide on the keywords recommendations through
+their usage data reduces the amount of bike-shedding discussions that currently
+exist when categories are decided upon with very little up-front usage data.
+
+
+# Drawbacks
+
+- As mentioned above, the chance for having duplicates is increased since not
+everyone will use the list of recommended keywords. This can be countered to
+some degree with aliases for recommended keywords, but can't be prevented
+completely.
+
+- The main drawback of changing how we categorize crates is the change itself.
+Existing users and crate authors are used to having both concepts available,
+and the existing tooling like IDE autocomplete of categories assumes the same
+thing. It will take a bit of time and educating people to adjust the ecosystem
+to the new concept.
+
+
+# Implementation
+
+The following implementation steps are planned for the crates.io side:
+
+- [ ] Add a publicly visible list of recommended keywords based on existing
+keyword usage on crates.io and the list of existing categories.
+- [ ] Add a publish-time deprecation warning for crates that are published with
+only categories but no keywords.
+- [ ] Remove categories from the crates.io frontend:
+  - Crates inside category pages
+  - Categories list
+  - Category slugs list
+  - Crate details page sidebar
+  - Support for `category:foo` in search bar
+- [ ] Remove categories from the crates.io backend:
+  - `/crates?category=foo` returns an empty list
+  - `/crates/foo` returns an empty list of `categories`
+  - `/category_slugs` returns an empty list
+  - `/categories` returns an empty list
+  - `/categories/foo` returns `404 Not Found`
+  - `/crates/new` stops saving categories for new uploads
+- [ ] Remove categories from that crates.io database (incl. the `ltree`
+extension).
+- [ ] Add a publish-time deprecation warning for all crates that are published
+with categories.
+
+
+# Requirements for recommended keywords
+
+The following basic rules should be followed when promoting keywords to the list
+of recommended keywords.
+
+1. The keyword does not violate the [Code of Conduct](https://www.rust-lang.org/policies/code-of-conduct)
+of the Rust project. (Such keywords are not allowed anyway, but the crates.io
+team is not currently monitoring new additions to the list of keywords and will
+only act on this when users report violations)
+
+2. The keyword is already used by substantial amount of crates from separate
+authors. (Intentionally vague to account for the growth of the package registry)
+
+3. The keyword serves a purpose that is not covered by any other recommended
+keyword.
+
+
+# Unresolved questions
+
+- Should the `categories` field be deprecated in cargo too?
+  - Are third-party registries using categories?
+  - How much usage of cargo is with crates.io vs. third party registries?
+  - Would users be confused if it was only deprecated on crates.io but not in cargo?
```

Comment:
imo if we move forward with this RFC, we should deprecate it and should be in an adjoining RFC that get approved together.  We can transition out `categories` on an Edition boundary, having `cargo fix --edition` do the merge in `Cargo.toml`.

Reply:
I think that moving the true "removal" of categories (in the sense that Cargo no longer sends them) to an edition boundary would be okay, although I'm not quite sure how editions would work here since registries aren't really controlled by editions, and crates.io also isn't controlled by editions.

In terms of splitting the RFC: which parts exactly do you think would be better suited to an adjacent RFC? The main reason for combining the two is that, if the plan is to coordinate merging both at the same time, there doesn't seem to be much benefit to splitting them, but I'm also not 100% sure what split you're thinking of.

Regardless, I think that the deprecation of categories should at least be followed with an increase in the number of allowed keywords at the same time, to ensure that users have the ability to adequately tag their crates. We could potentially discuss some of the other features separately, though.